File: code/erlang/src/ParallelProgramming/others/NotifyTest.java
package com.ericsson.concurrent;

//http://www.blogjava.net/zhangwei217245/archive/2010/04/07/316526.html
public class NotifyTest {
	private String flag[] = {"true"};

	class NotifyThread extends Thread {
		public NotifyThread(String name) {
			super(name);
		}

		public void run() {
			try {
				sleep(3000);// Õ∆≥Ÿ3√Î÷”Õ®÷™
			} catch (InterruptedException e) {
				e.printStackTrace();
			}

			synchronized (flag) {
				flag[0] = "false";
				flag.notifyAll();
			}
		}
	};

	class WaitThread extends Thread {
		public WaitThread(String name) {
			super(name);
		}

		public void run() {
			synchronized (flag) {
			while (flag[0] != "false") {
				System.out.println(getName() + " begin waiting!");
				long waitTime = System.currentTimeMillis();
				try {
					flag.wait();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
				waitTime = System.currentTimeMillis() - waitTime;
				System.out.println("wait time :" + waitTime);
			}
			}
			System.out.println(getName() + " end waiting!");

		}
	}

	public static void main(String[] args) throws InterruptedException {
		System.out.println("Main Thread Run!");	
		NotifyTest test = new NotifyTest();
		NotifyThread notifyThread = test.new NotifyThread("notify01");
		WaitThread waitThread01 = test.new WaitThread("waiter01");
		WaitThread waitThread02 = test.new WaitThread("waiter02");
		WaitThread waitThread03 = test.new WaitThread("waiter03");
		notifyThread.start();
		waitThread01.start();
		waitThread02.start();
		waitThread03.start();
	}

}

File: code/erlang/src/ParallelProgramming/others/NotifyTest.scala
package com.ericsson.concurrent

import scala.actors._
import scala.actors.Actor._

object NotifyTest {

  class NotifyThread(val name: String, val pids:Actor*) extends Actor {
    def act() {
      Thread.sleep(3000)
      pids.foreach(pid => pid ! "go")
    }
  }

  class WaitThread(val name: String) extends Actor {
    def act() {
      println(name + " begin waiting!")
      var waitTime = System.currentTimeMillis;
      receive {
        case "go" => waitTime = System.currentTimeMillis - waitTime
			      println("wait time :" + waitTime)
			      println(name + " end waiting!")
        case  _ => println("unexpected")
      }
      
    }
  }

  def main(args: Array[String]) {
    println("Main Thread Run!");
    val waitThread01 = new WaitThread("waiter01")
    val waitThread02 = new WaitThread("waiter02")
    val waitThread03 = new WaitThread("waiter03")
    val notifyThread = new NotifyThread("notify01",waitThread01, waitThread02, waitThread03)
    waitThread01.start()
    waitThread02.start()
    waitThread03.start()
    notifyThread.start()

  }

}